File: src/pipeline/subscriber-matcher.ts
import type { CellTower, GeoZone } from '../types/tower.js';

/**
 * Subscriber-matching contract for modules 03/04 — requirement #3/#4.
 *
 * Modules 03/04 are PLAN.md-only: they genuinely depend on C-DOT's subscriber
 * database, which is not connected yet. The pipeline detects this by asking
 * this registry — a module is "available" only when it has actually registered
 * a real matcher. Nothing is registered today, so the pipeline always halts at
 * the subscriber-matching stage with an explicit, visible status rather than
 * inventing a subscriber list.
 *
 * When C-DOT provides the DB and modules 03/04 ship, they call
 * `registerSubscriberMatcher(...)` with a real implementation and the pipeline
 * continues past this stage with no further changes.
 */

export interface SubscriberMatch {
  /** Tower ID whose footprint produced these recipients. */
  towerId: string;
  /** Real MSISDNs inside that tower's coverage. */
  msisdns: string[];
}

export interface SubscriberMatchContext {
  alertId: string;
  capIdentifier: string;
  /**
   * The alert zone (the drawn polygon / circles). Tower-based matchers ignore
   * it; the real C-DOT dump matcher uses it for direct point-in-polygon
   * matching against the subscriber geometry column.
   */
  zone?: GeoZone;
}

export interface SubscriberMatcher {
  readonly name: string;
  matchSubscribers(
    towers: readonly CellTower[],
    ctx: SubscriberMatchContext,
  ): Promise<SubscriberMatch[]>;
}

let registered: SubscriberMatcher | null = null;

/** Modules 03/04 call this when a real implementation ships. */
export function registerSubscriberMatcher(matcher: SubscriberMatcher): void {
  registered = matcher;
}

/** The currently-registered real matcher, or null while modules 03/04 are not built. */
export function getSubscriberMatcher(): SubscriberMatcher | null {
  return registered;
}

/** Unregister the matcher (test isolation; never called by the running app). */
export function resetSubscriberMatcher(): void {
  registered = null;
}

File: src/main/java/com/turant/subscriber/SubscriberMatcher.java
package com.turant.subscriber;

import com.turant.types.tower.CellTower;

import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Subscriber matcher contract - modules 03/04.
 * 
 * Given cell towers from module 02, returns all subscribers attached to those towers.
 * 
 * Migrated from TypeScript pipeline/subscriber-matcher.ts
 */
public interface SubscriberMatcher {
    
    /**
     * Stable name for logs/audit.
     */
    String getName();
    
    /**
     * Match subscribers to cell towers.
     *
     * @param towers List of cell towers from module 02
     * @param context Alert context (for logging and tracing)
     * @return List of subscriber matches per tower
     */
    CompletableFuture<List<SubscriberMatch>> matchSubscribers(
        List<CellTower> towers,
        MatchContext context
    );
    
    /**
     * Match result: tower → list of MSISDNs.
     */
    record SubscriberMatch(
        String towerId,
        List<String> msisdns
    ) {}
    
    /**
     * Context for subscriber matching.
     */
    record MatchContext(
        String alertId,
        String capIdentifier
    ) {}
}
